codegen/src/generator/mod.rs

```diff
@@ -1,59 +1,114 @@
 pub trait GenerateProtocol {
-    fn generate_methods(&self, service: &Service) -> String;
 
-    fn generate_prelude(&self, service: &Service) -> String;
+    /// Generate the various `use` statements required by the module generatedfor this service
+    fn generate_prelude(&self, writer: &mut FileWriter, service: &Service) -> IoResult;
+
+    /// Generate a method for each `Operation` in the `Service` to execute that method remotely
+    ///
+    /// The method generated by this method are inserted into an enclosing `impl FooClient {}` block
+    fn generate_methods(&self, writer: &mut FileWriter, service: &Service) -> IoResult;
 
-    fn generate_struct_attributes(&self, struct_name: &str) -> String;
+    /// Add any attributes that should decorate the struct for the given type (typically `Debug`, `Clone`, etc.)
+    fn generate_struct_attributes(&self,
+                                  struct_name: &str,
+                                  serialized: bool,
+                                  deserialized: bool)
+                                  -> String;
+
+    /// If necessary, generate a serializer for the specified type
+    fn generate_serializer(&self, _name: &str, _shape: &Shape, _service: &Service) -> Option<String> {
+        None
+    }
 
-    fn generate_support_types(&self,
-                              _name: &str,
-                              _shape: &Shape,
-                              _service: &Service)
-                              -> Option<String> {
+    /// If necessary, generate a deserializer for the specified type
+    fn generate_deserializer(&self, _name: &str, _shape: &Shape, _service: &Service) -> Option<String> {
         None
     }
 
+    /// Return the type used by this protocol for timestamps
     fn timestamp_type(&self) -> &'static str;
 }
 
-pub fn generate_source(service: &Service) -> String {
+/// Given a botocore `Service` object, determine its protocol and use the appropriate
+/// code generator to write code directly to a file at the specified path
+///
+/// # Panics
+/// If the service metadata indicates a protocol for which a code generator does not exist
+pub fn generate_source(service: &Service, output_path: &Path) -> IoResult {
+    let output_file = File::create(output_path).expect(&format!(
+        "Couldn't open file for writing: {:?}",
+        output_path,
+    ));
+
+    let mut writer = BufWriter::new(output_file);
+
     match &service.metadata.protocol[..] {
-        "json" => generate(service, JsonGenerator, JsonErrorTypes),
-        "ec2" => generate(service, Ec2Generator, XmlErrorTypes),
-        "query" => generate(service, QueryGenerator, XmlErrorTypes),
-        "rest-json" => generate(service, RestJsonGenerator, JsonErrorTypes),
-        "rest-xml" => generate(service, RestXmlGenerator, XmlErrorTypes),
+        "json" => generate(&mut writer, service, JsonGenerator, JsonErrorTypes),
+        "query" | "ec2" => generate(&mut writer, service, QueryGenerator, XmlErrorTypes),
```

Comment:
I'm going to document [why ec2 gets a special protocol](https://github.com/boto/botocore/blob/dff99fdf2666accf6b448aef7f03fe3d66dd38fa/botocore/serialize.py#L259-L266).